File: Week_1/pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

// Prototypes
int get_number_len(long number);

const long MAX_CARD_NUMBER = 999999999999999999;
const long MIN_CARD_NUMBER = 1000000000000;

const long DECIM_13_MIN = 1000000000000;
const long DECIM_13_MAX = 9999999999999;
const long DECIM_15_MIN = 100000000000000;
const long DECIM_15_MAX = 999999999999999;
const long DECIM_16_MIN = 1000000000000000;
const long DECIM_16_MAX = 9999999999999999;


// **********************************
// enter card number
// **********************************
long get_card_number(void)
{
    long card_number;

    do
    {
        card_number = get_long("Number: ");

    } while (card_number < 0 );

    return card_number;
}

// ************************************************
// summarize the multiplied 2 every other digits
// ************************************************
int get_sum_every_other(long number)
{
    int card_number_length = get_number_len(number);
    int sum = 0;

    for (long i = card_number_length - 1, j = 10; i >=0; i--, j *= 100)
    {

        int digit = ((number / j) % 10) * 2;

        // check if product has 2 digits and devide it for two digits
        if (digit < 10)
        {
            sum += digit;
        }
        else
        {
            int last_digit = digit % 10;
            int first_digit = 1;
            sum += (first_digit + last_digit);
        }

    }

    return sum;
}

// **********************************
// sum every first digits
// **********************************
int get_sum_every_first(long number)
{
    int sum = 0;

    for (long i = 0, j = 1; i < 8; i++, j *= 100)
    {
        sum += (number / j) % 10;
    }

    return sum;
}

// **********************************
// check if card valid
// **********************************
bool is_card_valid(long card_number)
{

    // check numbers quantity
    if (card_number > MAX_CARD_NUMBER || card_number < MIN_CARD_NUMBER) return false;

   // check if card is valid (by Luhn algorithm)
    int sum_every_other = get_sum_every_other(card_number);
    printf("%i ", sum_every_other);

    int sum_every_first = get_sum_every_first(card_number);
    printf("%i\n", sum_every_first);

    int sum = sum_every_other + sum_every_first;

    if (sum % 10 == 0) return true;
    else return false;

}

// **********************************
// calculate length of card number
// **********************************
int get_number_len(long number)
{
    if (number >= DECIM_13_MIN && number <= DECIM_13_MAX)      return 13;
    else if (number >= DECIM_15_MIN && number <= DECIM_15_MAX) return 15;
    else if (number >= DECIM_16_MIN && number <= DECIM_16_MAX) return 16;
    else return 0;
}

// **************************************
// extract first 2 digits from card number
// ***************************************
long get_first_two_digit(long number)
{
    int number_length =  get_number_len(number);

    // not valid card namber length
    if (number_length == 0) return 0;

    int first_two_digit;

    // calculate first 2 digits
    if (number_length == 13) first_two_digit = (number / (DECIM_13_MIN / 10));
    if (number_length == 15) first_two_digit = (number / (DECIM_15_MIN / 10));
    if (number_length == 16) first_two_digit = (number / (DECIM_16_MIN / 10));

    return first_two_digit;
}

// *************************************
// return car type based on firs 2 digits
// *************************************
string get_card_type(int first_two_digit)
{
    if (first_two_digit == 34 || first_two_digit == 37) return "AMEX";
    if (first_two_digit >= 51 && first_two_digit <= 55) return "MASTERCARD";
    if (first_two_digit >= 40 && first_two_digit <= 49) return "VISA";

    return "INVALID";
}

int main(void)
{
    // get card number from input
    long card_number = get_card_number();

    // get 2 fist digits of the card number
    int first_two_digit = get_first_two_digit(card_number);
    printf("%i\n", first_two_digit);

    // get card type
    string card_type = get_card_type(first_two_digit);
    printf("%s\n", card_type);

    // check if card is valid by Luhns algotrthm
    if (is_card_valid(card_number)) printf("%s\n", card_type);
    else printf ("%s\n", "INVALID");

}

File: Week_6/pset6/credit.py
# validate credit card using luhn algorithm
from functools import reduce

MAX_CARD_NUMBER = 9999_9999_9999_9999_99
MIN_CARD_NUMBER = 1000_0000_0000_0


def input_credit_card():
    """get card number from user input in the appropriate format"""

    try:
        number = int(input("Number: "))
    except ValueError:
        return None

    if MIN_CARD_NUMBER <= number <= MAX_CARD_NUMBER:
        return number
    else:
        return -1


def get_first_two_digit(card_number):
    """extract first 2 digits from card number"""

    return int(str(card_number)[:2])


def get_card_type(first_two_digit):
    """return card type based on firs 2 digits"""

    if first_two_digit == 34 or first_two_digit == 37:
        return "AMEX"

    if 51 <= first_two_digit <= 55:
        return "MASTERCARD"

    if 40 <= first_two_digit <= 49:
        return "VISA"

    return "INVALID"


def sum_every_other(digits_list):
    """return sum of every other digits in the `list` multiplied by 2"""
    every_other_list = [2 * x for x in digits_list[-2::-2]]
    plained_list = list(
        map(lambda x: [x] if x < 10 else convert_to_list(x),  every_other_list))
    result = reduce(lambda x, acc: x + acc, plained_list, [])
    return sum(result)


def convert_to_list(number):
    """helper convert digits in the number to list of digits"""
    return [int(x) for x in str(number)]


def get_sum_every_first(digits_list):
    """return sum of every first digits starting from the end"""
    return sum(digits_list[-1::-2])


def check_valid_card(number):
    """return `true` if if the total modulo 10 is congruent to 0 """
    return True if number % 10 == 0 else False


def main():

    # get correct credit card number from the user input
    card_number = None
    while card_number == None:
        card_number = input_credit_card()
    
    if card_number == -1:
        print("INVALID")
        return


    # validate card
    digits_list = convert_to_list(card_number)
    valid = check_valid_card(sum_every_other(digits_list) + get_sum_every_first(digits_list))
    if not valid:
        print("INVALID")
        return

    # get 2 fist digits of the card number
    first_two_digit = get_first_two_digit(card_number)
    
    # print card type
    print(get_card_type(first_two_digit))


if __name__ == "__main__":
    main()
